components/script/dom/htmlmediaelement.rs

```diff
@@ -1350,6 +1368,35 @@ impl HTMLMediaElementMethods for HTMLMediaElement {
         self.paused.get()
     }
 
+    /// https://html.spec.whatwg.org/multipage/#dom-media-defaultplaybackrate
+    fn DefaultPlaybackRate(&self) -> Finite<f64> {
+        Finite::wrap(self.defaultPlaybackRate.get())
+    }
+
+    /// https://html.spec.whatwg.org/multipage/#dom-media-defaultplaybackrate
+    fn SetDefaultPlaybackRate(&self, value: Finite<f64>) {
+        if *value != self.defaultPlaybackRate.get() {
+            self.defaultPlaybackRate.set(*value);
+            self.queue_ratechange_event();
+        }
+    }
+
+    /// https://html.spec.whatwg.org/multipage/#dom-media-playbackrate
+    fn PlaybackRate(&self) -> Finite<f64> {
+        Finite::wrap(self.playbackRate.get())
+    }
+
+    /// https://html.spec.whatwg.org/multipage/#dom-media-playbackrate
+    fn SetPlaybackRate(&self, value: Finite<f64>) {
+        if *value != self.playbackRate.get() {
```

Comment:
Also according to the spec, we should only change the rate if we are [potentially playing](https://html.spec.whatwg.org/multipage/media.html#potentially-playing)

Reply:
I looked at how Gecko [does](https://github.com/mozilla/gecko-dev/blob/47edbd91c43db6229cf32d1fc4bae9b325b9e2d0/dom/html/HTMLMediaElement.cpp#L5772) this check and noticed that it misses some steps in the process. I am not really sure how to go about implementing it, could you, please, give me some guidelines?

Reply:
Yeah, we don't have all the required pieces to completely implement all the checks, but we can at least leave the corresponding placeholders:

- We depend  on #22348 to get the `Ended` method.

- I suspect we depend on the MediaSession API to be able to implement the `is_paused_for_user_interaction` method.

- We certainly depend on #22314 to implement `is_paused_for_in_band_content`. Let's just leave the placeholders for now.

I would do something like:

```rust
fn is_potentially_playing(&self) -> bool {
    !self.paused.get() &&
    // FIXME: We need https://github.com/servo/servo/pull/22348
    //              to know whether playback has ended or not
    // !self.Ended() &&
    self.error.get().is_none() &&
    !self.is_blocked_media_element()
}

fn is_blocked_media_element(&self) -> bool {
    self.readyState.get() <= ReadyState::HaveCurrentData ||
    self.is_paused_for_user_interaction() ||
    self.is_paused_for_in_band_content()
}

fn is_paused_for_user_interaction(&self) -> bool {
    // FIXME: we will likely be able to fill this placeholder once (if) we
    //             implement the MediaSession API.
    false
}

fn is_paused_for_in_band_content(&self) -> bool {
     // FIXME: we will likely be able to fill this placeholder once (if) we implement
     //             https://github.com/servo/servo/issues/22314
    false
}
```